### DISCUSSION TITLE
Are the source file paths the same across the generated gcda files and gcno symlinks?

### DISCUSSION THREAD

**Discussion by joefradley** [linux-test-project/lcov#224] [Q&A] [answered]
I’m looking at fixing up the paths of the *.gcda and *.gcno symlink files populated into debugfs when gcov is enabled in the kernel. The paths aren’t accurate due to sandboxing done at build time. We have generated mapping files to help with this process.

My question is, are all the following <dev_src_path> values going to be the same?:
 1. <debugfs_path>/gcov/**<dev_src_path>**.gcda
 2. <debugfs_path>/gcov/**<dev_src_path>**.gcno → **<dev_src_path>**.gcno

The reason I ask is it would be convenient if I didn’t have to evaluate the values of the symlinks.

**Comment by henry2cox** [comment, marked as answer]
I'm not sure that I understand the question.
All else being equal:  the .gcno and .gcda files do get dumped into the same location (next to the .o's) - but I'm not sure if kernel builds and kernel modules are the same as regular user code.
If someone used the GCOV_PREFIX (and GCOV_PREFX_STRIP) environment variables:  then the .gcda files are going to go elsewhere (into the prefix directory, with paths reflecting the number of levels that got stripped).
Some time back, I had added some 'capture' features to make it easier to capture data from split builds.  See the geninfo man page for '--build-directory'.

If the question is about the source paths that are found in the captured .info file:  then those are closely related to how your code was built.  If the paths are not what you like - then you can use the "--substitute" option to change them (or simply use sed to get equivalent results).
The lcov tools are very, very dumb - and basically require _exact_ pathname matches to decide that a particular file is the same or not.

Hope this helped...but please clarify the question if possible.
Henry

**Reply by joefradley** [reply]
(oof sorry this wasn't a proper reply at first)

Thanks for the response. My question is more about the structure of the created debugfs gcov directory in order to fix it up before `lcov` is called. Perhaps using `--substitute` and other lcov features can be used instead of this pre-processing.  I vaguely remember GCOV_PREFIX and GCOV_PREFIX_STRIP, I can't recall if there was a reason they couldn't be used for my situation. I'll have to revisit them. It could be just what I'm looking for.

> All else being equal: the .gcno and .gcda files do get dumped into the same location (next to the .o's) - but I'm not sure if kernel builds and kernel modules are the same as regular user code.

Anecdotally, I feel this is the case for the kernel and modules.  I just wanted to make sure there aren't situations where it isn't.

**Comment by oberpar** [comment]
> My question is, are all the following <dev_src_path> values going to be the same?:
> 
>     1. <debugfs_path>/gcov/**<dev_src_path>**.gcda
> 
>     2. <debugfs_path>/gcov/**<dev_src_path>**.gcno → **<dev_src_path>**.gcno
> 
> 
> The reason I ask is it would be convenient if I didn’t have to evaluate the values of the symlinks.

As Henry mentioned, `.gcda` and `.gcno` files are typically located in the same directory. This is also true for gcov-kernel. The link target will point to **<dev_src_path>** unless you're using `O=` or `M=` parameters during kernel compilation.
